src/librustdoc/passes/collect_intra_doc_links/early.rs

```diff
@@ -1,75 +1,148 @@
-// crates the TyCtxt sees and the resolver sees (because the resolver could load more crates
-// after escaping). Hopefully `IntraLinkCrateLoader` gets all the crates we need ...
-crate fn load_intra_link_crates(resolver: Resolver, krate: &ast::Crate) -> Resolver {
-    let mut loader = IntraLinkCrateLoader { current_mod: CRATE_DEF_ID, resolver };
-    // `walk_crate` doesn't visit the crate itself for some reason.
+
+crate fn load_intra_link_crates(
+    resolver: &mut Resolver<'_>,
+    krate: &ast::Crate,
+) -> DefIdMap<DefIdSet> {
+    let mut loader = IntraLinkCrateLoader {
+        resolver,
+        current_mod: CRATE_DEF_ID,
+        visited_mods: Default::default(),
+        traits_in_scope: Default::default(),
+    };
+
+    // Overridden `visit_item` below doesn't apply to the crate root,
+    // so we have to visit its attributes and exports separately.
     loader.load_links_in_attrs(&krate.attrs, krate.span);
-    ast::visit::walk_crate(&mut loader, krate);
-    loader.resolver
+    loader.process_exports(CRATE_DEF_ID.to_def_id());
+    visit::walk_crate(&mut loader, krate);
+    loader.add_foreign_traits_in_scope();
+
+    loader.traits_in_scope
 }
 
-struct IntraLinkCrateLoader {
+struct IntraLinkCrateLoader<'r, 'ra> {
+    resolver: &'r mut Resolver<'ra>,
     current_mod: LocalDefId,
-    resolver: Rc<RefCell<interface::BoxedResolver>>,
+    visited_mods: DefIdSet,
+    traits_in_scope: DefIdMap<DefIdSet>,
 }
 
-impl IntraLinkCrateLoader {
-    fn load_links_in_attrs(&mut self, attrs: &[ast::Attribute], span: Span) {
-        use crate::html::markdown::markdown_links;
-        use crate::passes::collect_intra_doc_links::preprocess_link;
+impl IntraLinkCrateLoader<'_, '_> {
+    fn add_traits_in_scope(&mut self, module_id: DefId) {
+        if let Entry::Vacant(entry) = self.traits_in_scope.entry(module_id) {
+            if self.resolver.can_get_module(module_id) {
+                let parent_scope =
+                    ParentScope::module(self.resolver.get_module(module_id), self.resolver);
+                entry.insert(
+                    self.resolver
+                        .traits_in_scope(None, &parent_scope, SyntaxContext::root(), None)
+                        .into_iter()
+                        .map(|candidate| candidate.def_id)
+                        .collect(),
+                );
+            }
+        }
+    }
+
+    /// Add traits in scope for links in impls collected by the `collect-intra-doc-links` pass.
+    /// That pass filters impls using type-based information, but we don't yet have such
+    /// information here, so we just conservatively calculate traits in scope for *all* modules
+    /// having impls in them.
+    fn add_foreign_traits_in_scope(&mut self) {
+        for cnum in self.resolver.cstore().crates_untracked() {
+            // For some reason rustdoc also wants to know traits in scope for every crate root.
+            self.add_traits_in_scope(cnum.as_def_id());
```

Comment:
Hmm, that's surprising to me - the only way the module should be from an external crate is if the item being documented was inlined from that crate. So it shouldn't need *each* crate, only the ones where an item is inlined.

In practice it doesn't really matter, because `IntraLinkCrateLoader` doesn't know which items are inlined. But it would be nice for the comment to be correct.

```suggestion
            // Add the traits in the crate root, since an item defined in the root might be inlined to the current crate.
            self.add_traits_in_scope(cnum.as_def_id());
```

Reply:
Items cannot be inlined without exports (`use`), right?
After adding the export processing code that should cover inlining I tried to remove this line, and majority of tests in the `rustdoc` test suite failed.
I didn't figure out why this happens, but it seems to be not inlining.
Maybe it's `crate::foo` paths in impls collected by `collect-intra-doc-links`, I remember seeing some logic about that, need to investigate.

Reply:
>  Maybe it's crate::foo paths in impls collected by collect-intra-doc-links, I remember seeing some logic about that, need to investigate.

Oh, this is probably https://github.com/rust-lang/rust/issues/78696. Fixing it is waiting on https://github.com/rust-lang/rust/pull/78797#issuecomment-731538913 IIRC.

Reply:
Added a comment referencing https://github.com/rust-lang/rust/issues/78696.